caffe2/contrib/tensorrt/tensorrt_op_trt.cc

```diff
@@ -54,14 +57,51 @@ TensorRTOp::TensorRTOp(const OperatorDef& operator_def, Workspace* ws)
   {
     auto engine_string =
         OperatorBase::GetSingleArgument<std::string>("serialized_engine", "");
-    CAFFE_ENFORCE(!engine_string.empty(), "Empty serialized TensorRT engine!");
-    auto trt_runtime = tensorrt::TrtObject(nvinfer1::createInferRuntime(logger_));
-    // TODO(support trt plugin factory)
-    trt_engine_ = tensorrt::TrtObject(trt_runtime->deserializeCudaEngine(
-        engine_string.data(), engine_string.size(), nullptr));
+    if (!engine_string.empty()) {
+      auto trt_runtime =
+          tensorrt::TrtObject(nvinfer1::createInferRuntime(logger_));
+      // TODO(support trt plugin factory)
+      trt_engine_ = tensorrt::TrtObject(trt_runtime->deserializeCudaEngine(
+          engine_string.data(), engine_string.size(), nullptr));
+    } else {
+      auto onnx_model_str =
+          OperatorBase::GetSingleArgument<std::string>("onnx_model", "");
+      CAFFE_ENFORCE(!onnx_model_str.empty(), "onnx_model cannot be empty");
+      auto debug_builder = OperatorBase::GetSingleArgument<int>("debug_builder", 0);
+      auto max_workspace_size = OperatorBase::GetSingleArgument<int>(
+          "max_workspace_size", 1024 * 1024 * 2);
+
+      // Pull the weights from workspace and assembly it back to the onnx model,
+      // notice that since we may have rewritten the net, we need to map the
+      // weight names
+      auto initializers = OperatorBase::GetRepeatedArgument<std::string>("initializers");
+      CAFFE_ENFORCE_EQ(
+          initializers.size() % 2, 0, "initializers should come in pairs");
+      std::unordered_set<std::string> initializer_set;
+      std::unordered_map<std::string, std::string> input_mapping;
+      for (auto it = initializers.begin(); it != initializers.end(); ++it)  {
+        auto key = *it++;
+        input_mapping.emplace(key, *it);
+        initializer_set.emplace(key);
+      }
+      Workspace mapped_ws(ws, input_mapping);
```

Comment:
Does this create a copy of all blobs in the workspace? If so, we need to find a way to avoid it.

Reply:
It doesn't. It creates a view towards the parent Workspace with a name mapping. Exactly what I wanted. :)